src/py42/clients/archive.py

```diff
@@ -184,16 +193,19 @@ def stream_to_device(
             overwrite_existing_files,
         )
 
-    def _select_backup_set_id(self, device_guid, destination_guid):
+    def _select_backup_set_id(self, device_guid, destination_guid, backup_set_id):
         backup_sets = self.get_backup_sets(device_guid, destination_guid)["backupSets"]
-        if not backup_sets:
-            return None
-        for backup_set in backup_sets:
-            backup_set_id = backup_set["backupSetId"]
-            if str(backup_set_id) == _DEFAULT_BACKUP_SET_ID:
-                return backup_set_id
-
-        return backup_sets[0]["backupSetId"]
+        backup_set_ids = [bs["backupSetId"] for bs in backup_sets]
+        if backup_set_id:
+            if backup_set_id not in backup_set_ids:
+                raise Py42Error(f"{backup_set_id=} not found in device backup sets: {backup_sets}")
+            return backup_set_id
+        # id=1 is the "default" backup set, use this if it exists and no explicit id
+        # is provided by user
+        elif "1" in backup_set_ids:
+            return "1"
+        else:
+            return backup_set_ids[0]
```

Comment:
Do we still need to catch this case of no backup_sets being returned? Looks like it may KeyError in the `else` statement

Reply:
I'm pretty sure it's not possible for a device to have _no_ backup sets at all, and if we do return `None` the restore requests would fail anyway as it's a required value. 

I changed it to raise an exception if no sets are found, to fail fast and explicitly if we do ever hit this case.